WTF Memory management functions for Haiku

pulkomandy opened the pull request:
#### 99dbf190711012921585003f644a6f8993d8b71c
<pre>
WTF Memory management functions for Haiku

Reviewed by Adrian Perez de Castro.

Add needed support to get memory size and usage on Haiku.
Existing code is not suitable because of use of getrusage (not all
returned fields are in POSIX), /proc filesystem, or other nonstandard
code.

No new tests: no impact on other ports.
* Source/WTF/wtf/FastMalloc.cpp:
(WTF::fastMallocStatistics): Implement getting RAM usage for Haiku.
* Source/WTF/wtf/PlatformHaiku.cmake: Add new files to the build.
* Source/WTF/wtf/haiku/CurrentProcessMemoryStatus.cpp: Added.
(WTF::currentProcessMemoryStatus): Get memory usage for Haiku.
* Source/WTF/wtf/haiku/CurrentProcessMemoryStatus.h: Added.
* Source/WTF/wtf/haiku/MemoryFootprintHaiku.cpp: Added.
(WTF::memoryFootprint): Get memory usage for Haiku.
* Source/WTF/wtf/unix/MemoryPressureHandlerUnix.cpp:
(WTF::processMemoryUsage): Wire access to memory usage in a way similar to Linux.

Canonical link: <a href="https://commits.webkit.org/304097@main">https://commits.webkit.org/304097@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3979369a7a5de443fc76d2d083c572ad6bcac125

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134574 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7032 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45809 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142099 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86531 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7634 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6890 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102859 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/70140 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137521 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5342 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120622 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83657 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5200 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2818 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2701 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126622 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38758 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144794 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/133082 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6707 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39340 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111258 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6781 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5627 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111542 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28284 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5036 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116892 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/60581 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6758 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35085 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165996 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6570 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70341 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43391 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6805 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6681 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WTF/wtf/haiku/CurrentProcessMemoryStatus.h:
```diff
@@ -0,0 +1,45 @@
+namespace WTF {
+
+struct ProcessMemoryStatus {
+    size_t size { 0 };
+    size_t resident { 0 };
+    size_t shared { 0 };
+    size_t text { 0 };
+    size_t lib { 0 };
+    size_t data { 0 };
+    size_t dt { 0 };
```
Small nit: IMO it does not make sense to zero-initialize all the members here, given that they will be reset inside `currentProcessMemoryStatus()` anyway.

pulkomandy replied:
This file is copied from the Linux implementation: https://github.com/WebKit/WebKit/blob/3979369a7a5de443fc76d2d083c572ad6bcac125/Source/WTF/wtf/linux/CurrentProcessMemoryStatus.h 

But the corresponding cpp file is different, and indeed the Haiku implementation cannot fail and will always reset the structure, while the Linux one can fail. So the initialization here seems unnecessary for Haiku.

aperezdc replied:
Good point in them having the same members. Maybe we should look into sharing the definition, and guard it with `OS(LINUX) || OS(HAIKU)`.

Also, searching for uses of `ProcessMemoryStatus`, it is never heap-allocated, so we could mark it with `WTF_FORBID_HEAP_ALLOCATION` 🤔 

Anyway, I think these are things that can be done later on as cleanup patches. Let's land this as-is.

aperezdc reviewed (approved):
LGTM, with a small nit (see below); I am not an expert using the Haiku/Be APIs but looking at [OS.h](https://github.com/haiku/haiku/blob/master/headers/os/kernel/OS.h) usage seems reasonable 👌🏼

pulkomandy committed:
WTF Memory management functions for Haiku

Reviewed by Adrian Perez de Castro.

Add needed support to get memory size and usage on Haiku.
Existing code is not suitable because of use of getrusage (not all
returned fields are in POSIX), /proc filesystem, or other nonstandard
code.

No new tests: no impact on other ports.
* Source/WTF/wtf/FastMalloc.cpp:
(WTF::fastMallocStatistics): Implement getting RAM usage for Haiku.
* Source/WTF/wtf/PlatformHaiku.cmake: Add new files to the build.
* Source/WTF/wtf/haiku/CurrentProcessMemoryStatus.cpp: Added.
(WTF::currentProcessMemoryStatus): Get memory usage for Haiku.
* Source/WTF/wtf/haiku/CurrentProcessMemoryStatus.h: Added.
* Source/WTF/wtf/haiku/MemoryFootprintHaiku.cpp: Added.
(WTF::memoryFootprint): Get memory usage for Haiku.
* Source/WTF/wtf/unix/MemoryPressureHandlerUnix.cpp:
(WTF::processMemoryUsage): Wire access to memory usage in a way similar to Linux.

Canonical link: https://commits.webkit.org/304097@main
